spmd/compiler/fusion.py

```diff
@@ -162,22 +214,448 @@ def _scan_graph_for_fusion_elements(
     return element_list
 
 
-def run_comm_fusion(gm: fx.GraphModule) -> bool:
+def _copy_fe_to_buffer(
+    gi: GraphInfo, gm: fx.GraphModule, in_fe_list: list[FusionElement]
+) -> None:
+    """first half of fusion - move desired items to buffer and create graph"""
+    buffer_node = gi.global_buffer_node
+    buffer_size = gi.global_buffer_size
+
+    copy_list = in_fe_list
+
+    num_fusion_elements = len(copy_list)
+
+    def copy_to_buffer(
+        buffer: torch.Tensor, tensor_list: List[torch.Tensor]
+    ) -> torch.Tensor:
+        offset = 0
+        for t in tensor_list:
+            size = t.numel()
+            buffer[offset : offset + size] = t.view(-1)
+            offset += size
+        return buffer
+
+    # setup dummy vars
+    buffer = None
+    if gi.tracing_buffer is None:
+        buffer = torch.empty(buffer_size)
+        gi.tracing_buffer = buffer
+    elif gi.tracing_buffer:
+        buffer = buffer = torch.empty(buffer_size)
+
+    tlist = []
+    for item in copy_list:
+        a = torch.zeros((item.shape[0], item.shape[1]))  # type: ignore
+        tlist.append(a)
+
+    load_gm = make_fx(copy_to_buffer)(buffer, tlist)
+
+    subnodemap = create_graph_node_map(load_gm)
+
+    # update load loop to use main graph items
+    fn_list = []
+    pl_list = []
+    for node in load_gm.graph.nodes:
+        if node.op == OP.PLACEHOLDER:
+            pl_list.append(node)
+        elif node.op == OP.CALL_FUNCTION:
+            fn_list.append(node)
+
+    # create placeholder remapping
+    pl_map: Dict[fx.Node, fx.Node] = {}
+    pl_map[pl_list[0]] = gi.global_buffer_node  # type: ignore
+
+    for i in range(num_fusion_elements):
+        pl_map[pl_list[i + 1]] = in_fe_list[i].grad_tensor_node  # type: ignore
+
+    insert_node = in_fe_list[-1].comm_node
+
+    def remap_copy_args(in_node: fx.Node) -> fx.Node:
+        out_node = in_node
+        if in_node in pl_map:
+            out_node = pl_map[in_node]  # type: ignore
+        elif in_node in value_remap:
+            out_node = value_remap[in_node]
+        return out_node
+
+    value_remap = {}
```

Comment:
should this be defined before the above function?

Reply:
it technically works as is b/c the earlier use is an embedded function that is only invoked after this...but agree it's confusing and cleaner to have it above. 

Moved it up and also added in typing for it:
value_remap: Dict[fx.Node, fx.Node] = {}